src/coreclr/src/jit/morph.cpp

```diff
@@ -10134,13 +10133,23 @@ GenTree* Compiler::fgMorphBlkNode(GenTree* tree, bool isDest)
     }
     GenTree* blkSrc = blkNode->Addr();
     assert(blkSrc != nullptr);
-    if (!blkNode->TypeIs(TYP_STRUCT) && blkSrc->OperIs(GT_ADDR) && blkSrc->gtGetOp1()->OperIs(GT_LCL_VAR))
+    if (blkSrc->OperIs(GT_ADDR) && blkSrc->gtGetOp1()->OperIs(GT_LCL_VAR))
     {
         GenTreeLclVarCommon* lclVarNode = blkSrc->gtGetOp1()->AsLclVarCommon();
-        if ((genTypeSize(blkNode) != genTypeSize(lclVarNode)) || (!isDest && !varTypeIsStruct(lclVarNode)))
+        if ((blkNode->TypeGet() != TYP_STRUCT) &&
+            ((genTypeSize(blkNode) != genTypeSize(lclVarNode)) || (!isDest && !varTypeIsStruct(lclVarNode))))
         {
             lvaSetVarDoNotEnregister(lclVarNode->GetLclNum() DEBUG_ARG(DNER_VMNeedsStackAddr));
         }
+        else
+        {
+            unsigned lclSize = lclVarNode->TypeIs(TYP_STRUCT) ? lvaGetDesc(lclVarNode)->lvExactSize
```

Comment:
Is that change related to multi-reg or is it an independent optimization? 
It looks like it replaces `BLK(ADDR(lvlVar))` with `lclVar` if they have the same size, do I understand that correctly?
`lclVarNode->TypeGet() == tree->TypeGet()` is very weak for struct type, do you want to check `CORINFO_CLASS_HANDLE` instead? With the current checks, I think it could be broken with a few unsafe casts.